docs/design/resource-states.md

```diff
@@ -0,0 +1,213 @@
+2. The State of the resource in Azure. What fields are set in Azure? There are often defaults and other values which, while you may not have specified them in the `spec` have been applied on the server side and we want to show to you.
+
+This document is focused on the first type of Status defined above.
+
+## Current state of ASO
+
+### ASO v1 (handcrafted)
+
+The handcrafted ASO resources have a 3-tuple on the `Status` field which describes the state of the resource:
+```go
+	Provisioning       bool                `json:"provisioning,omitempty"`
+	Provisioned        bool                `json:"provisioned,omitempty"`
+	FailedProvisioning bool                `json:"failedProvisioning,omitempty"`
+```
+
+
+There are also additional fields for conveying additional information, including the details of any error as well as information about the Resource ID of the resource
+```go
+	State              string              `json:"state,omitempty"`
+	Message            string              `json:"message,omitempty"`
+	ResourceId         string              `json:"resourceId,omitempty"`
+```
+
+#### Problems with this approach
+* `Provisioning`, `Provisioned` and `FailedProvisioning` are mutually exclusive but the structure of the `status` doesn't convey that well. Additionally the fact that there are multiple
+  fields means it's easier for bugs to creep in where we set one but didn't clear another.
+* The `State` field is not widely used by all resources, and seems to conflict with `Provisioning`/`Provisioned`/`FailedProvisioning`. It's not clear when you should look at one or the other.
+
+### ASO v2 (autogenerated)
+The autogenerated resources don't current have anything in their `Status` which conveys the Status _of the operator_. For the initial alpha releases, these fields are instead set as annotations on the object:
+```go
+	DeploymentIDAnnotation   = "deployment-id.infra.azure.com"
+	DeploymentNameAnnotation = "deployment-name.infra.azure.com"
+	ResourceStateAnnotation  = "resource-state.infra.azure.com"
+	ResourceErrorAnnotation  = "resource-error.infra.azure.com"
+```
+
+This was done mostly to avoid the additional work to code-generate the correct fields in each resources `status` and set them. Since it's possible that the whole `status` can get lost, some of the things may need to stay 
+as annotations (`deploymentId` possibly), while others are obviously status and can be derived again if lost. We took the temporary approach of putting it all in annotations while we learned more about what the best practices were.
+
+#### Problems with this apporach
+* The `ResourceStateAnnotation` is reusing the same enum to describe state as the `armclient` uses for deployments. This is problematic because that only supports `Succeeded`, `Failed`, `Deleting`, and `Accepted`.
+  `Succeeded`, `Failed`, and `Deleting` make sense in the context of the operator but `Accepted` doesn't really. There's no state for `InProgress` or `Reconciling`.
```

Comment:
also Failed and Deleting don't talk about the resource they talk about the deployment; the deployment can be Deleting but the resource still exists.

Reply:
Yeah, I'll update the doc to call that out more clearly.